### PR TITLE
Implement draft RPC API.

### PR DESCRIPTION
**Pull request by tmpfs**
Remove old Callback mechanism.

* Remove obsolete Callback API
* Remove FuncCall and RPC
* Update README
* Rename set_rpc_handler() to set_handler()
* Use shared rpc_proxy() function for platform consistency
* Improve handling of promise cleanup
* Update README with RPC API info.
* Panic if webview handler is set after window creation.
* Improve rpc_proxy() logic, try to ensure any corresponding promise is always removed.
* Remove FuncCall wrapper.

### CODE REVIEW

**Comment by tmpfs** [comment]
@wusyong, this version has a tidy history. Would appreciate it if we could get this merged :+1: 

I tried and failed to fix the MacOS issue and hope you have more success. I tried using `OnceCell` to get an `&'static` reference to the `RpcHandler` but could not as it is not thread safe. It appears that ultimately it boils down to:

```
= help: within `[closure@src/application/general.rs:370:52: 380:10]`, the trait `Sync` is not implemented for `*mut winit::platform_impl::platform::observer::CFRunLoopSource`
```

Which is all the stuff inside the `WindowProxy` -> `ApplicationProxy` chain which I know very little about. Any idea how we can get the `RpcHandler` into `did_receive()` safely?

**Comment by wusyong** [comment]
We probably need to declare another objc class and it's full of hacks. I can take another day to work it. But for now that's just make it work on linux.

**Comment by tmpfs** [comment]
Thanks @wusyong, it works on Linux do you have a Windows machine to test on? It might work there too!

**Comment by wusyong** [comment]
Sure. Would you mind I cherry pick this comment to `feat/rpc`? I wish I could work on it right away.

**Comment by tmpfs** [comment]
Sure thing, let's put it in `feat/rpc` and merge it back into the main branch once we have all three platforms :+1: Let me know if there is anything else I can do to help. Maybe a tracking issue for progress on this feature?

**Comment by wusyong** [comment]
We can keep the discussion on #66.
Currently I am thinking about making `Callback` in the application module work with rpc.
I will list more in the issue later.

**Comment by tmpfs** [comment]
> Currently I am thinking about making Callback in the application module work with rpc.

Oh no, we may have misunderstood each other. I have removed `Callback` as I don't see the benefit it brings to have a separate Rust closure for each Javascript function. The new model is a single Rust closure which tests on the `method` of `RpcRequest`.

The benefit of the old callback model was that you could just call a vanilla Javascript function as I understood it. This can now be achieved like this:

```javascript
function someCallback() {
  return window.rpc.call('someRpcMethod', Array.prototype.slice.call(arguments, 0));
}

someCallback("param").then((res) => console.log(res));
```

What is still not done is a function that would set these *callback aliases* automatically. That would be the `set_rpc_aliases()` function I mentioned before.

**Comment by wusyong** [comment]
I see. I do agree with this and it looks a lot cleaner.
Btw it turns out it's hell of chore to cherry pick 😅. Could you let me just add them manually?

**Comment by tmpfs** [comment]
Hi @wusyong, sorry I didn't realize you have been working in the `feat/rpc` branch. I took a look at the commits and there are lots of conflicts with that I have done here. Particularly with regard to `CALLBACKS` and `Callback`.

What I attempted to do was be fully backwards compatible with the first PR that you merged and then when you said that you liked the idea of cleaning up `FuncCall` and `Callback` this PR was created that *completely removes any concept of callbacks* with the *exception of Javascript-defined callback functions* that wrap the RPC API.

Is it possible for me to take any important changes from your `feat/rpc` branch and merge them with this PR? Not sure which changes are important. Please advise.

**Comment by wusyong** [comment]
Callback has to be here is because Tauri will need it. But I believe there's a way to build it on top of the RPC handler.
Let me try to merge your branch into `feat/rpc` and I'll list what should be done afterward in #66
We can remove callback which's fine, but it needs an alternative interface that could accomplish the same functionality.

**Comment by wusyong** [comment]
Sorry I deleted the previous comment thinking it's not good for you to develop further.
I think let's just keep working on this branch and finish all the changes you would like to address (ie. RPC alias).
Forget about `feat/rpc`, I'll update them afterward once this PR is merged.

**Comment by tmpfs** [comment]
Sure @wusyong, I'll do the callback aliases stuff and hope that it works out.

I looked at the [tauri code](https://github.com/tauri-apps/tauri/blob/adac93b9f10284e10bc58b436bb91284ed2cd132/tauri/src/app/utils.rs#L183) that configures the `Callback` and all they would need to do is change it to `Box<RpcHandler>` and call `set_handler()` when the application is set up. But without knowing the code well it is hard to say how much effort it would be.

Maybe we can get some people working on the main Tauri code to comment on how much this change will affect them?

**Comment by wusyong** [comment]
Yeah we'll have a discussion about this.
Another thing I would like to address is maybe moves the handler to parameters of `Applicaton::create_window`.
I'm not a fan of requiring the correct order of calling methods (which we suffered a lot from original webview)

**Comment by tmpfs** [comment]
> I'm not a fan of requiring the correct order of calling methods 

Agreed @wusyong . I did add a `panic!` for out of order calling but it's still a bit rubbish.

So then we remove `set_handler()` and `RpcHandler` is passed as the second argument to `add_window_with_configs()`? Replacing the old `Vec<Callback>`.

**Comment by wusyong** [comment]
yeah, if that's possible.

**Comment by tmpfs** [comment]
Sure thing @wusyong, that's done now and it feels more idiomatic :+1: 

However it does mean that for the *aliases* functionality that they would need to be passed in using the same mechanism so we probably want a wrapper type:

```rust
pub struct RpcConfig {
  handler: RpcHandler,
  callbacks: Vec<String>
}
```

Then we pass `Option<RpcConfig>` to `add_window_with_configs()` instead of `Option<RpcHandler>`; is that ok with you?

**Comment by wusyong** [comment]
Current `Option<RpcHandler>` is cleaner imho. Let me check with other devs if we can actually replace callbacks first.

File: README.md
Comment (wusyong) [inline README.md:49]: I left a comment in #39 that wish we could add descriptions to examples. Maybe you could start a `README.md` in `examples/`

File: examples/rpc.rs
Comment (wusyong) [inline examples/rpc.rs:40]: This comment could be removed right?

File: src/application/mod.rs
Comment (wusyong) [inline src/application/mod.rs:46]: Let's kill these once and for all 🔪

File: src/application/mod.rs
Comment (wusyong) [inline src/application/mod.rs:296]: It's not included in the error type anymore, so Debug trait is not required.

File: src/platform/linux.rs
Comment (wusyong) [inline src/platform/linux.rs:11]: You could remove it if not needed.

File: src/webview.rs
Comment (wusyong) [inline src/webview.rs:37]: yeah you could remove it since we just needed it for callbacks before.

**Review by wusyong — CHANGES_REQUESTED** [review]
Great news! We could get rid of callbacks entirely! Just fix the following nits and I think it's good to go!

**Comment by tmpfs** [comment]
Thanks @wusyong, excited to see this land! I have made those small changes you mentioned but I have a couple more points I would like to make:

1) We throw out the idea of *callback aliases* and just add documentation on how to do it in Javascript
2) I read the Tauri code some more and they do async stuff in the existing `Callback` code and we need to support that use case.

I have a good idea how we can support async flows it it just a small bit of code and some documentation. I will do it now and ask you to take a look, thanks :+1:

**Comment by tmpfs** [comment]
Hi @wusyong, I think this can be merged if we can answer the following questions:

1. Does it work on Windows?
2. Do we have a clear path to get `RpcHandler` into `did_receive()` for MacOS support?

Hoping you are able to work out the MacOS issues! Cheers :+1:

File: examples/multiwindow.rs
Comment (wusyong) [inline examples/multiwindow.rs:19]:
```suggestion
        url: Some(format!("data:text/html,{}", html)),
```
It won't be rendered on windows if it's encoded. Looks like the web engine will do this for us.
Could you remove this dependency?

Comment (tmpfs) [reply examples/multiwindow.rs]: Sure thing, that's done! Strange they accept it un-encoded, I assume the browser is decoding before passing to the webview, all good :+1:

**Comment by tmpfs** [comment]
Thanks for all your help @wusyong, shall we get this merged and then over to you for the platform-specific stuff?

**Review by wusyong — APPROVED** [review]
Sure, really appreciate the help!

**Pull request merged**
